activation/activation.go

```diff
@@ -472,74 +527,55 @@ func (b *Builder) Coinbase() types.Address {
 	return b.coinbaseAccount
 }
 
-func (b *Builder) loadChallenge() (*types.NIPostChallenge, error) {
-	nipost, err := LoadNipostChallenge(b.nipostBuilder.DataDir())
-	if err != nil {
-		return nil, err
-	}
-	if nipost.PublishEpoch < b.currentEpoch() {
-		b.log.With().Info("atx nipost challenge is stale - discarding it",
-			log.Stringer("publish_epoch", nipost.PublishEpoch),
-			log.Stringer("current_epoch", b.currentEpoch()),
-		)
-		if err = b.discardChallenge(); err != nil {
-			return nil, fmt.Errorf("%w: atx nipost challenge is stale", err)
-		}
-		return nil, errors.New("atx nipost challenge is stale")
-	}
-	return nipost, nil
-}
-
 // PublishActivationTx attempts to publish an atx, it returns an error if an atx cannot be created.
 func (b *Builder) PublishActivationTx(ctx context.Context) error {
-	logger := b.log.WithContext(ctx)
-
-	challenge, err := b.loadChallenge()
+	challenge, err := b.buildNIPostChallenge(ctx)
 	if err != nil {
-		if !errors.Is(err, os.ErrNotExist) {
-			logger.With().Warning("failed to load atx challenge", log.Err(err))
-		}
-		logger.With().Info("building new atx challenge",
-			log.Stringer("current_epoch", b.currentEpoch()),
-		)
-		challenge, err = b.buildNIPostChallenge(ctx)
-		if err != nil {
-			return err
-		}
+		return err
 	}
 
-	logger.With().Info("atx challenge is ready",
-		log.Stringer("current_epoch", b.currentEpoch()),
-		log.Stringer("publish_epoch", challenge.PublishEpoch),
-		log.Stringer("target_epoch", challenge.TargetEpoch()),
+	b.log.Info("atx challenge is ready",
+		zap.Stringer("current_epoch", b.layerClock.CurrentLayer().GetEpoch()),
+		zap.Stringer("publish_epoch", challenge.PublishEpoch),
+		zap.Stringer("target_epoch", challenge.TargetEpoch()),
 	)
+	ctx, cancel := context.WithDeadline(ctx, b.layerClock.LayerToTime((challenge.TargetEpoch()).FirstLayer()))
+	defer cancel()
+	atx, err := b.createAtx(ctx, challenge)
+	if err != nil {
+		return fmt.Errorf("create ATX: %w", err)
+	}
 
-	if b.pendingATX == nil {
-		var err error
-		ctx, cancel := context.WithDeadline(ctx, b.layerClock.LayerToTime((challenge.TargetEpoch()).FirstLayer()))
-		defer cancel()
-		b.pendingATX, err = b.createAtx(ctx, challenge)
-		if err != nil {
-			return fmt.Errorf("create ATX: %w", err)
+	for {
+		size, err := b.broadcast(ctx, atx)
+		if err == nil {
+			b.log.Info("atx published", zap.Inline(atx), zap.Int("size", size))
+			break
+		}
+		if errors.Is(err, context.DeadlineExceeded) {
```

Comment:
This is actually a hack to make some tests easier. If the context expires it will be caught by the `select` statement after this check.

Looking again at it I think I can adapt the test to not require this hack any more.

Reply:
I updated the test so that the special handling of `context.DeadlineExceeded` isn't necessary any more.